contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,440 @@
+        burned_ = _d - _D(xWad, yWad);
+        _d -= burned_;
+
+        uint256[] memory fees = new uint256[](2);
+        emit RemoveLiquidity(msg.sender, amounts, _d);
+    }
+
+    /// @notice Swaps token from for an amount of token to
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    /// @param minDy The minimum amount of token out in token decimals
+    /// @param deadline The deadline before which swap must be executed
+    // TODO: test w protocol fees
+    function swap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx,
+        uint256 minDy,
+        uint256 deadline
+    )
+        external
+        nonReentrant
+        onlyToken(tokenIndexFrom)
+        onlyToken(tokenIndexTo)
+        deadlineCheck(deadline)
+        hasQuote
+        returns (uint256 dy_)
+    {
+        require(tokenIndexFrom != tokenIndexTo, "invalid token swap");
+
+        // get current token in balance in pool
+        address tokenIn = tokenIndexFrom == 0 ? token0 : token1;
+        uint256 bal = IERC20(tokenIn).balanceOf(address(this));
+
+        // transfer in tokens and update dx (in case of transfer fees)
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), dx);
+        dx = IERC20(tokenIn).balanceOf(address(this)) - bal;
+
+        // calculate amount out from swap
+        // @dev returns zero if amount out exceeds pool balance
+        uint256 dyAdminFee;
+        (dy_, , dyAdminFee) = _calculateSwap(tokenIndexFrom, tokenIndexTo, dx);
+        require(dy_ > 0, "dy > pool balance");
+        require(dy_ >= minDy, "dy < minDy");
+
+        if (tokenIndexTo == 0) {
+            // update admin fee reserves
+            protocolFees0 += dyAdminFee;
+
+            // transfer dy out to swapper
+            address tokenOut = token0;
+            IERC20(tokenOut).safeTransfer(msg.sender, dy_);
+        } else {
+            // update admin fee reserves
+            protocolFees1 += dyAdminFee;
+
+            // transfer dy out to swapper
+            address tokenOut = token1;
+            IERC20(tokenOut).safeTransfer(msg.sender, dy_);
+        }
+
+        emit TokenSwap(msg.sender, dx, dy_, tokenIndexFrom, tokenIndexTo);
+    }
+
+    /// @notice Transfers protocol fees out
+    /// @param recipient The recipient address of the aggregated admin fees
+    // TODO: test
+    function skim(address recipient) external onlyFactory {
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = protocolFees0;
+        amounts[1] = protocolFees1;
+
+        // clear protocol fees
+        protocolFees0 = 0;
+        protocolFees1 = 0;
+
+        // send out to factory owner
+        IERC20(token0).safeTransfer(recipient, amounts[0]);
+        IERC20(token1).safeTransfer(recipient, amounts[1]);
+
+        emit Skim(msg.sender, amounts);
```

Comment:
I think emitting `recipient` instead of the factory feels more natural (factory is immutable, but its owner could potentially change over time).